src/SourceGenerator.Foundations.Contracts/ISgfInitializationContext.cs

```diff
@@ -0,0 +1,34 @@
+﻿using System;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using SGF.Diagnostics;
+
+namespace SGF;
+
+public interface ISgfInitializationContext
+{
+    SyntaxValueProvider SyntaxProvider { get; }
+    IncrementalValueProvider<Compilation> CompilationProvider { get; }
+    IncrementalValueProvider<ParseOptions> ParseOptionsProvider { get; }
+    IncrementalValuesProvider<AdditionalText> AdditionalTextsProvider { get; }
+    IncrementalValueProvider<AnalyzerConfigOptionsProvider> AnalyzerConfigOptionsProvider { get; }
+    IncrementalValuesProvider<MetadataReference> MetadataReferencesProvider { get; }
+    /// <summary>
+    /// Gets the underlying <see cref="ILogger"/> 
+    /// </summary>
+    /// <remarks>Provided as a convenience accessor to support advanced scenarios. Generally not needed but no harm if used.</remarks>
+    ILogger Logger { get; }
+    /// <summary>
+    /// Gets the original <see cref="IncrementalGeneratorInitializationContext"/> 
+    /// </summary>
+    /// <remarks>Provided as a convenience accessor to support advanced scenarios. Generally should not be directly
+    /// interacted with as it may produce side effects in the sgf generators. Use at your own risk.</remarks>
+    IncrementalGeneratorInitializationContext Context { get; }
```

Comment:
I think it would be better to hide this from the interface and instead add it to the implementation then update the generators to take in  `ISgfInitializationContext` instead. 

This
```cs
public abstract void OnInitialize(SgfInitializationContext context);
```
would become 
```cs
public abstract void OnInitialize(ISgfInitializationContext context);
```
The reason being is that people don't often read the documentation and learn from intelisense. If we expose the property they might use it and when they do they loose all the exception handling that SGF provides. There are for sure going to be use cases where you need it but for folks who really need it can read the docs or find out how to get it will a little digging.